oxide-vpc/tests/integration_tests.rs

```diff
@@ -2218,204 +2217,356 @@ fn gateway_router_advert_reply() {
+    assert_eq!(ip6.src, na.src_ip);
+    assert_eq!(ip6.dst, na.dst_ip);
+    assert_eq!(ip6.proto, Protocol::ICMPv6);
 
+    // Validate the details of the Neighbor Advertisement itself.
     let rdr = reply.get_body_rdr();
     let reply_body = rdr.copy_remaining();
     let reply_pkt = Icmpv6Packet::new_checked(&reply_body).unwrap();
     let mut csum = CsumCapab::ignored();
     csum.icmpv6 = smoltcp::phy::Checksum::Rx;
-    let reply_icmp =
-        Icmpv6Repr::parse(&src.into(), &dst.into(), &reply_pkt, &csum).unwrap();
-    match reply_icmp {
-        Icmpv6Repr::Ndisc(NdiscRepr::NeighborAdvert {
-            flags,
-            target_addr,
+    let reply_icmp = Icmpv6Repr::parse(
+        &IpAddress::Ipv6(ip6.src.into()),
+        &IpAddress::Ipv6(ip6.dst.into()),
+        &reply_pkt,
+        &csum,
+    )
+    .unwrap();
+    if let Icmpv6Repr::Ndisc(NdiscRepr::NeighborAdvert {
+        flags,
+        target_addr,
+        lladdr,
+    }) = reply_icmp
+    {
+        assert_eq!(flags, na.flags);
+        assert_eq!(target_addr, na.target_addr.into());
+        assert_eq!(
             lladdr,
-        }) => {
-            assert_eq!(
-                flags,
-                NdiscNeighborFlags::ROUTER
-                    | NdiscNeighborFlags::SOLICITED
-                    | NdiscNeighborFlags::OVERRIDE
-            );
-            assert_eq!(
-                target_addr,
-                Ipv6Addr::from_eui64(&g1_cfg.gateway_mac).into()
-            );
-            assert_eq!(
-                lladdr.expect("Expected a Link-Layer Address option"),
-                RawHardwareAddress::from_bytes(&g1_cfg.gateway_mac.bytes())
-            );
-        }
-        other => {
-            panic!(
-                "Expected an ICMPv6 Neighbor Advertisement, found {:?}",
-                other
-            );
-        }
-    };
+            na.lladdr.map(|x| RawHardwareAddress::from_bytes(&x.bytes()))
+        );
+    } else {
+        panic!(
+            "Expected an ICMPv6 Neighbor Advertisement, found {:?}",
+            reply_icmp
+        );
+    }
+}
+
+// Ensure that we either Drop a Neighbor Solicitation, or generate a Neighbor
+// Advertisement with the right data, based on our defined test cases.
+#[test]
+fn test_gateway_neighbor_advert_reply() {
+    let g1_cfg = g1_cfg();
+    let v2p = Arc::new(Virt2Phys::new());
+    let mut ameta = ActionMeta::new();
+    let mut g1 = oxide_net_setup("g1_port", &g1_cfg, v2p.clone());
+    g1.port.start();
+    set_state!(g1, PortState::Running);
+    let mut pcap = PcapBuilder::new("gateway_neighbor_advert_reply.pcap");
+
+    let data = generate_solicit_test_data(&g1_cfg);
+    for d in data.into_iter().skip(1).take(1) {
+        let mut pkt = generate_neighbor_solicitation(&d.ns);
+        pcap.add_pkt(&pkt);
+        let res = g1.port.process(Out, &mut pkt, &mut ameta);
+        // Hack to deal appease borrow checker, since `res` isn't `Copy`.
+        let res_dbg = format!("{:?}", &res);
```

Comment:
The problem was that when you tried to format it directly in `panic!` it was borrowing the `res` from this lexical scope (don't ask me why I don't know). If you explicitly match in the panic arm with `(res, _)` you don't need to do this.

Reply:
I believe I tried that, without success. IIRC, then we can't pass the `hp` variable we match in the second arm to the core validation function `validate_hairpin_advert`. If we pass `hp` by reference there, then we can't call `hp.parse()` inside the function, since that takes `mut self`. But if we pass `hp` by value, then we have a partial move out of `res`, meaning we can't put it in the `panic!` message. There are other ways around this, but it's all kinda hacky and fine for the tests.